### PR TITLE
clippy run

### PR DESCRIPTION
**Pull request by tshepang**

### CODE REVIEW

File: src/cargo/core/package_id_spec.rs [hunk 1]
```diff
@@ -154,10 +155,10 @@ impl PackageIdSpec {
                     ids: Vec<&PackageId>,
                     spec: &PackageIdSpec) {
             let mut version_cnt = HashMap::new();
-            for id in ids.iter() {
+            for id in &ids {
```
Comment (alexcrichton) [inline src/cargo/core/package_id_spec.rs]: I personally prefer `.iter()` instead of `&`, can this stay the same?

File: src/cargo/core/package.rs [hunk 2]
```diff
@@ -145,7 +145,7 @@ impl<'cfg> PackageSet<'cfg> {
         }
     }
 
-    pub fn package_ids<'a>(&'a self) -> Box<Iterator<Item=&'a PackageId> + 'a> {
+    pub fn package_ids<'a>(&'a self) -> Box<Iterator<Item=&PackageId> + 'a> {
```
Comment (alexcrichton) [inline src/cargo/core/package.rs]: This seems.. odd?

File: src/cargo/core/registry.rs [hunk 3]
```diff
@@ -166,9 +166,9 @@ impl<'cfg> PackageRegistry<'cfg> {
 
     pub fn register_lock(&mut self, id: PackageId, deps: Vec<PackageId>) {
         let sub_map = self.locked.entry(id.source_id().clone())
-                                 .or_insert(HashMap::new());
+                                 .or_insert_with(HashMap::new);
```
Comment (alexcrichton) [inline src/cargo/core/registry.rs]: This seems kinda unclear to me, can it stay the way it was before? (same below)

File: src/cargo/core/resolver/encode.rs [hunk 4]
```diff
@@ -86,9 +86,8 @@ impl EncodableResolve {
                 for edge in deps.iter() {
                     let to_depend_on = try!(id2pkgid(edge));
                     let precise_pkgid =
-                        tmp.get(&to_depend_on)
-                           .map(|p| to_depend_on.with_precise(p.clone()))
-                           .unwrap_or(to_depend_on.clone());
+                        tmp.get(&to_depend_on).map_or(to_depend_on.clone(),
+                                |p| to_depend_on.with_precise(p.clone()));
```
Comment (alexcrichton) [inline src/cargo/core/resolver/encode.rs]: I personally prefer map + unwrap_or, can this stay the same?

File: src/cargo/core/shell.rs [hunk 5]
```diff
@@ -122,12 +122,9 @@ impl MultiShell {
 
     pub fn set_color_config(&mut self, color: Option<&str>) -> CargoResult<()> {
         let cfg = match color {
-            Some("auto") => Auto,
+            Some("auto") | None => Auto,
             Some("always") => Always,
             Some("never") => Never,
-
-            None => Auto,
```
Comment (alexcrichton) [inline src/cargo/core/shell.rs]: I'd prefer this stayed a separate case to be clear as to what the default is

File: src/cargo/ops/cargo_new.rs [hunk 6]
```diff
@@ -378,8 +378,7 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
 
     let vcs = match (opts.version_control, cfg.version_control, in_existing_vcs_repo) {
         (None, None, false) => VersionControl::Git,
-        (None, Some(option), false) => option,
-        (Some(option), _, _) => option,
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs]: I'd prefer these stay separate to be clear what's being matched

File: src/cargo/ops/cargo_rustc/links.rs [hunk 7]
```diff
@@ -4,6 +4,7 @@ use core::PackageId;
 use util::CargoResult;
 use super::Unit;
 
+#[derive(Default)]
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/links.rs]: Why was this added?

File: src/cargo/sources/registry/mod.rs [hunk 8]
```diff
@@ -322,10 +322,10 @@ impl<'cfg> Registry for RegistrySource<'cfg> {
         // theory the registry is known to contain this version. If, however, we
         // come back with no summaries, then our registry may need to be
         // updated, so we fall back to performing a lazy update.
-        if dep.source_id().precise().is_some() && !self.updated {
-            if try!(self.index.query(dep)).is_empty() {
-                try!(self.do_update());
-            }
+        if dep.source_id().precise().is_some()
+        && !self.updated
+        && try!(self.index.query(dep)).is_empty() {
```
Comment (alexcrichton) [inline src/cargo/sources/registry/mod.rs]: This seems unreadable now?

File: src/cargo/util/hex.rs [hunk 9]
```diff
@@ -4,7 +4,7 @@ use rustc_serialize::hex::ToHex;
 
 pub fn to_hex(num: u64) -> String {
     [
-        (num >>  0) as u8,
+        (num      ) as u8,
```
Comment (alexcrichton) [inline src/cargo/util/hex.rs]: This was done for clarity, it should stay the same

File: src/cargo/util/paths.rs [hunk 10]
```diff
@@ -68,33 +68,33 @@ pub fn without_prefix<'a>(a: &'a Path, b: &'a Path) -> Option<&'a Path> {
 }
 
 pub fn read(path: &Path) -> CargoResult<String> {
-    (|| -> CargoResult<_> {
+    {
```
Comment (alexcrichton) [inline src/cargo/util/paths.rs]: This should not change, it's now semantically different

File: src/cargo/ops/cargo_rustc/links.rs (hunk 7 above)
Comment (tshepang) [inline src/cargo/ops/cargo_rustc/links.rs]: Was following https://github.com/Manishearth/rust-clippy/wiki#new_without_default

File: src/cargo/sources/registry/mod.rs (hunk 8 above)
Comment (tshepang) [inline src/cargo/sources/registry/mod.rs]: Is it the formatting making it not readable?

File: src/cargo/ops/cargo_rustc/links.rs (hunk 7 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/links.rs]: Yes but this isn't needed, nor is it a public API, so there's no need to add it

File: src/cargo/sources/registry/mod.rs (hunk 8 above)
Comment (alexcrichton) [inline src/cargo/sources/registry/mod.rs]: Yes

File: src/cargo/core/registry.rs (hunk 3 above)
Comment (killercup) [inline src/cargo/core/registry.rs]: I'd be interested to know whether the HashMap is always created in the old version, or if this can be optimized away.

(Maybe you can also to `.or_insert_with(|| HashMap::new())` if @alexcrichton just doesn't want to have this curried?)

Comment (tshepang) [inline src/cargo/core/registry.rs]: According to https://github.com/Manishearth/rust-clippy/wiki#or_fun_call, the old version will indeed always be called.

Comment (alexcrichton) [inline src/cargo/core/registry.rs]: Yes but `or_insert_with(HashMap::new)` in this case looks less readable to me, so I'd prefer to leave as-is.

**Comment by tshepang** [comment]
pushed empty commit by mistake

**Comment by tshepang** [comment]
review addressed here: https://github.com/rust-lang/cargo/pull/3048
